Improves processing of multiple grains during dnatracing

ns-rse opened the pull request:
Closes #289
Closes #294 

This removes processing of a whole image and instead makes the unit of processing in dnatracing a single grain. Processing multiple grains is done via a helper function `trace_image()` which calls `trace_grain()` the later is a thin wrapper to the `traceDna()` class.

When cropping the image and masks for processing they are padded twice compared to the scikit-image determined bounding box.

1. The original image/mask are paded by a user configurable amount. This is so that heights which will be used when height-based tracing is introduced are available outside of the region.
2. These images are then padded further to allow some grains that are close to the edge/border of images to be processed.

Currently the value for both steps of padding is the same and is set to `20`, other values may be more appropriate.

Two bugs :lady_beetle: were identified during this work and have (hopefully) been corrected. See commits 236750b2 and 2a79c4ff for further details (thanks @SylviaWhittle for finding and fixing one of those).

Summary of changes.

* Introduces `scikit-image` methods for skeletonisation (still a Work In Progress as they are not pruned).
* Updates tests for test_tracing_methods to work with dnatracing now loops have been removed.
* Uses loops to pass over all grains in tracing because can't run parallel process from within a parallelised process.
* Code left in place for running in parallel when we get round to modularising running TopoStats properly.
* Adds a new function for preparing arrays (`prep_arrays()`) along with tests.
* Removes the class TraceStats completely no longer required
* Removes medial as a method of skeletonisation
* Some (but not all) docstrings added (most will get added during subsequent refactoring I hope).
* Updates target regression test files.
* Tidies up and removes redundant tests.

There is still considerable refactoring to do including...

* Removal of plotting from `dnaTrace` class and adding functionality to `plottingfuncs.Images()` class to overlay or just plot traces on their own.
* Check and improve splining, there are some notes suggesting this doesn't work on some shapes.
* Prune skeletons returned by scikit-image methods.
* Move some vectorised methods on other branches into `main` (mainly the method for counting number of neighbours from `ns-rse/183b-dnatracing-refactor`.
* Lots of more tests.

SylviaWhittle reviewed (approved):
The code makes sense, and it appears to do everything as before. So this is a tentative approval from me!

I have also tried it on several non-minicircle images, including some non-bio images. A couple did flag errors occuring during dna tracing, however I removed the `try...except` from around dna tracing and confirmed that the same errors happen in main too, so nothing to worry about. (Some division by zero errors in `get_splined_traces` etc).

This PR really lays the solid foundation for further improvements in dna molecule tracing, and we really appreciate the amount of hard work and testing that went into this @ns-rse . The code is so clear and follow-able now, reading through it during the review was a pleasant and intuitive experience (ignoring the legacy code that we are planning on stripping out later on). 😄 

I've flagged a few things that aren't an issue, but I just wanted to make sure you were aware of.

I've also suggested a couple naming clarifications, but again these are completely optional and aren't really necessary.

I look forward to working on adding functionality to this now it isn't scary and cryptic!

Thanks again!

ns-rse commented on topostats/tracing/dnatracing.py:
```diff
@@ -758,117 +638,284 @@ def measure_contour_length(self):
+    cropped_images, cropped_masks = prep_arrays(image, grains_mask, pad_width)
+    n_grains = len(cropped_images)
+    LOGGER.info(f"[{filename}] : Calculating statistics for {n_grains} grains.")
+    grain = 0
```
Good point, gone with `n_grain` to be consistent with the `trace_grain()` argument.

ns-rse commented on topostats/tracing/dnatracing.py:
```diff
@@ -758,117 +638,284 @@ def measure_contour_length(self):
+#              the labelled regions. @ns-rse
+#
+#
+# if __name__ == "__main__":
```
Yep, plan is to add to the "swiss-army knife" implementation once your PR (#540) is merged. The code here (and related imports) work so felt it was useful to leave in place for when that is undertaken.

ns-rse commented on topostats/tracing/dnatracing.py:
```diff
@@ -81,644 +114,491 @@ def __init__(
         plt.show()
         plt.close()
 
     def saveTraceFigures(
```
I'd used it a few times whilst doing this work to check what images are like, as we've discussed plotting the traces should be an additional method to the `plottingfuncs.Image` class with the ability to overlay on different plots, whether that is the original image of filtered heights, cropped images, masks (full or cropped) or just the trace on its own.

We'll have to decide how to return and amalgamate the different objects and find out what people want saving (#595), e.g. should we include unordered traces, I suspect not but it may be desirable to have them available perhaps.

ns-rse commented on tests/test_io.py:
```diff
@@ -305,13 +305,14 @@ def test_get_out_path_attributeerror() -> None:
 
 
-def test_save_folder_grainstats(tmp_path: Path, minicircle_tracestats: pd.DataFrame) -> None:
+def test_save_folder_grainstats(tmp_path: Path) -> None:
```
I'm happy `pd.DataFrame.to_csv()` works as expected and writes an ASCII CSV, the test here checks the file is written to the correct place which is mostly what `io.save_folder_grainstats()` is doing.

ns-rse commented on topostats/default_config.yaml:
```diff
@@ -51,6 +51,9 @@ grainstats:
   min_skeleton_size: 10 # Minimum number of pixels in a skeleton for it to be retained.
+  skeletonisation_method: topostats # Options : zhang | lee | thin | topostats
+  pad_width: 20 # Cells to pad grains by when tracing
+#  cores: 1 # Number of cores to use for parallel processing
```
Yes, that' the plan once #540 is merged (not checked it might already) I'll get an entry point setup.

ns-rse commented:
> I have also tried it on several non-minicircle images, including some non-bio images. A couple did flag errors occuring during dna tracing, however I removed the try...except from around dna tracing and confirmed that the same errors happen in main too, so nothing to worry about. (Some division by zero errors in `get_splined_traces` etc).

Thanks for checking, we've a good candidate for the next investigation there (the broad `try: except:` makes debugging really awkward.


> This PR really lays the solid foundation for further improvements in dna molecule tracing, and we really appreciate the amount of hard work and testing that went into this @ns-rse . The code is so clear and follow-able now, reading through it during the review was a pleasant and intuitive experience (ignoring the legacy code that we are planning on stripping out later on). smile

Very kind although I haven't substantially changed what is being done more the how, but it should make it easier to extend.

> I've flagged a few things that aren't an issue, but I just wanted to make sure you were aware of.
> 
> I've also suggested a couple naming clarifications, but again these are completely optional and aren't really necessary.

All perfectly sane and addressed.
